p4_symbolic/ir/ir.cc

```diff
@@ -0,0 +1,295 @@
+  // TODO(babman): Code duplication between this and TransformLValue.
+  //               This function will have more cases later when the
+  //               second todo is handled, but we still need to reduce
+  //               code duplication.
+  //               Difficulty: the type here is RValue instead of LValue.
+  //               Possible solution: create piece wise re-usable functions
+  //               that return FieldValue/Variable etc, cons: does not reduce
+  //               duplicates by much..
+  //               Will look at this later.
+  // TODO(babman): Support the remaining cases: literals and simple expressions.
+  if (bmv2_value.kind_case() != google::protobuf::Value::kStructValue) {
+    return absl::Status(absl::StatusCode::kInvalidArgument,
+                        "Left-hand of assignment is badly formatted!");
+  }
+
+  const google::protobuf::Struct &struct_value = bmv2_value.struct_value();
+  const std::string &type = struct_value.fields().at("type").string_value();
+  if (type == "field") {
+    const google::protobuf::ListValue &names =
+        struct_value.fields().at("value").list_value();
+
+    FieldValue *field_value = dst->mutable_field_value();
+    field_value->set_header_name(names.values(0).string_value());
+    field_value->set_field_name(names.values(1).string_value());
+  } else if (type == "runtime_data") {
+    int variable_index = struct_value.fields().at("value").number_value();
+
+    Variable *variable = dst->mutable_variable_value();
+    variable->set_name(variables[variable_index]);
+  } else {
+    return absl::Status(absl::StatusCode::kUnimplemented,
+                        "Unsupported expression in left-hand of assignment!");
+  }
+
+  return absl::OkStatus();
+}
+
+// Parsing and validating actions.
+absl::Status TransformAction(const bmv2::Action &bmv2_action,
+                             const pdpi::ir::IrActionDefinition &pdpi_action,
+                             Action *output) {
+  // Definition is copied form pdpi.
+  output->mutable_action_definition()->CopyFrom(pdpi_action);
+
+  // Implementation is extracted from bmv2.
+  ActionImplementation *action_impl = output->mutable_action_implementation();
+
+  // BMV2 format uses ints as ids for variables.
+  // We will replace the ids with the actual variable name.
+  std::vector<std::string> variable_map(bmv2_action.runtime_data_size());
+  for (int i = 0; i < bmv2_action.runtime_data_size(); i++) {
+    const bmv2::VariableDefinition variable = bmv2_action.runtime_data(i);
+    (*action_impl->mutable_variables())[variable.name()] = variable.bitwidth();
+    variable_map[i] = variable.name();
+  }
+
+  // Parse every statement in body.
+  // When encoutering a variable, look it up in the variable map.
+  for (int i = 0; i < bmv2_action.primitives_size(); i++) {
+    const google::protobuf::Struct &primitive = bmv2_action.primitives(i);
+    const std::string &op = primitive.fields().at("op").string_value();
```

Comment:
Still hoping we can rename "op" to "operation" (https://google.github.io/styleguide/cppguide.html#General_Naming_Rules)